src/main/java/org/ow2/proactive/catalog/rest/controller/BucketController.java

```diff
@@ -205,31 +206,39 @@ public List<BucketMetadata> list(
         contentType = contentType.filter(s -> !s.isEmpty());
         objectName = objectName.filter(s -> !s.isEmpty());
         if (sessionIdRequired) {
-            RestApiAccessResponse restApiAccessResponse = restApiAccessService.checkAccessBySessionIdForOwnerOrGroupAndThrowIfDeclined(sessionId,
-                                                                                                                                       ownerName);
+            AuthenticatedUser user = restApiAccessService.checkAccessBySessionIdForOwnerOrGroupAndThrowIfDeclined(sessionId,
+                                                                                                                  ownerName)
+                                                         .getAuthenticatedUser();
             log.debug("bucket list timer : validate session : " + (System.currentTimeMillis() - startTime) + " ms");
-            listBucket = bucketService.getBucketsByGroups(ownerName,
-                                                          kind,
-                                                          contentType,
-                                                          objectName,
-                                                          () -> restApiAccessResponse.getAuthenticatedUser()
-                                                                                     .getGroups());
-
-            AuthenticatedUser user = restApiAccessService.getUserFromSessionId(sessionId);
+            listBucket = bucketService.getBucketsByGroups(ownerName, kind, contentType, objectName, user::getGroups);
             listBucket.addAll(grantRightsService.getBucketsForUserByGrantsAndPriority(user));
-            List<BucketMetadata> res = new LinkedList<>();
-            for (BucketMetadata data : listBucket) {
-                String bucketGrantAccessType = grantRightsService.getResultingAccessTypeFromUserGrantsForBucketOperations(user,
-                                                                                                                          data.getName());
-                int objectCount = bucketGrantService.getTheNumberOfAccessibleObjectsInTheBucket(user, data);
-                BucketMetadata metadata = new BucketMetadata(data.getName(), data.getOwner(), objectCount);
-                metadata.setRights(bucketGrantAccessType);
-                if (!res.contains(metadata)) {
-                    res.add(metadata);
+
+            List<BucketGrantMetadata> allBucketsGrants = bucketGrantService.getUserAllBucketsGrants(user);
+            List<CatalogObjectGrantMetadata> allCatalogObjectsGrants = catalogObjectGrantService.getAllObjectsGrantsAssignedToAUser(user);
+
+            for (BucketMetadata bucket : listBucket) {
+                if (grantRightsService.isPublicBucket(bucket.getOwner())) {
+                    bucket.setRights(admin.name());
```

Comment:
I'm not sure if you should add the bucket object count here also (unless it's added by default when you get the bucket)

Reply:
Yes, it's already added when get the bucket.